### PR TITLE
Add a pallet to demonstrate calling into a contract

### PR DESCRIPTION
**Pull request by ascjones**
An foundational step towards https://github.com/paritytech/ink/issues/1674, developing an API for calling into a contract from the runtime.

This PR 
 - adds a new pallet `contract-caller` which allows calling into a flipper contract
 - uses the existing available API in `ink` to construct the call for the contract
 - adds the `flipper` contract with a trait and an e2e test which demonstrates it working: instantiates the contract, invokes the pallet, queries the modified state of the contract.
 - introduces a CI step to run the e2e test.

The next step (in progress) is to work and iterate on the `ink` API itself, which will involve an `ink` PR and a PR here to utilize the new API.

### CODE REVIEW

File: pallets/contract-caller/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+				contract.clone(),
+				value.into(),
+				weight,
+				storage_deposit_limit,
+				data,
+				pallet_contracts::DebugInfo::Skip,
+				pallet_contracts::CollectEvents::Skip,
+				pallet_contracts::Determinism::Enforced,
+			)
+			.result?;
```
Comment (ascjones) [inline pallets/contract-caller/src/lib.rs:81]: This is the code we want to simplify with a nice API in `ink!`

File: pallets/contract-caller/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	type AccountIdOf<Runtime> = <Runtime as frame_system::Config>::AccountId;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_contracts::Config {
+		/// The overarching runtime event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
```
[the remark this answers is not in the document]
Comment (ascjones) [reply pallets/contract-caller/src/lib.rs]: I don't think so. I've removed it.

File: pallets/contract-caller/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	use frame_system::pallet_prelude::*;
+	use ink::codegen::TraitCallBuilder;
+
+	type AccountIdOf<Runtime> = <Runtime as frame_system::Config>::AccountId;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_contracts::Config {
```
[the remark this answers is not in the document]
Comment (ascjones) [reply pallets/contract-caller/src/lib.rs]: The `#[pallet::config]` macro requires the explicit implementation of `frame_system::Config`

**Comment by jasl** [comment]
Do you think this PR can resolve my question? https://substrate.stackexchange.com/questions/10685/how-to-call-a-contract-from-pallet-with-specific-interface

I want to extend my pallet through the developers' uploaded contract,
for example,
a DAO as a service pallet, people can create DAOs, and the pallet would allow the DAO's owner to register "hooks" (for example `on_proposal_pass`) so they could have a restricted way to customize it.
So, I want those contracts that serve as a hook, acquired to conform to specific interfaces, and in the pallet, I want to call the typed interface instead of a selector.

**Comment by ascjones** [comment]
> Do you think this PR can resolve my question? https://substrate.stackexchange.com/questions/10685/how-to-call-a-contract-from-pallet-with-specific-interface
> 
> I want to extend my pallet through the developers' uploaded contract, for example, For an NFT marketplace pallet, I want to allow NFT collection owners to register a few "hooks" (such as `on_bought` and `on_sell`) so they can have a restricted way to extend fixed pallet business logic. So, I want those contracts that serve as a hook, acquired to conform to specific interfaces, and in the pallet, I want to call the typed interface instead of a selector.

Yes, you can use the technique I have used in this PR. I am currently working on a nicer API for this, but it is a few weeks away from being ready.

**Comment by jasl** [comment]
> > Do you think this PR can resolve my question? https://substrate.stackexchange.com/questions/10685/how-to-call-a-contract-from-pallet-with-specific-interface
> > I want to extend my pallet through the developers' uploaded contract, for example, For an NFT marketplace pallet, I want to allow NFT collection owners to register a few "hooks" (such as `on_bought` and `on_sell`) so they can have a restricted way to extend fixed pallet business logic. So, I want those contracts that serve as a hook, acquired to conform to specific interfaces, and in the pallet, I want to call the typed interface instead of a selector.
> 
> Yes, you can use the technique I have used in this PR. I am currently working on a nicer API for this, but it is a few weeks away from being ready.

That's awesome! Will try when it's landing

**Comment by ascjones** [comment]
Superseded by https://github.com/paritytech/ink/pull/2189, it is more appropriate to live in that repo since that is where the API will live.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/contracts-on-assethub-roadmap/9513/11

**Pull request closed without merging**